network/server.go

```diff
@@ -516,117 +776,146 @@ var (
+	close(s.closeCh)
 
 	return err
 }
 
-func (s *Server) watch(peerID peer.ID, dur time.Duration) error {
-	ch := make(chan error)
+// newProtoConnection opens up a new stream on the set protocol to the peer,
+// and returns a reference to the connection
+func (s *Server) newProtoConnection(protocol string, peerID peer.ID) (*rawGrpc.ClientConn, error) {
+	s.protocolsLock.Lock()
+	defer s.protocolsLock.Unlock()
 
-	s.joinWatchersLock.Lock()
-	if s.joinWatchers == nil {
-		s.joinWatchers = map[peer.ID]chan error{}
+	p, ok := s.protocols[protocol]
+	if !ok {
+		return nil, fmt.Errorf("protocol not found: %s", protocol)
 	}
 
-	s.joinWatchers[peerID] = ch
-	s.joinWatchersLock.Unlock()
-
-	select {
-	case <-time.After(dur):
-		s.joinWatchersLock.Lock()
-		delete(s.joinWatchers, peerID)
-		s.joinWatchersLock.Unlock()
-
-		return fmt.Errorf("timeout %s %s", s.host.ID(), peerID)
-	case err := <-ch:
-		return err
+	stream, err := s.NewStream(protocol, peerID)
+	if err != nil {
+		return nil, err
 	}
-}
-
-func (s *Server) runJoinWatcher() error {
-	return s.SubscribeFn(func(evnt *PeerEvent) {
-		switch evnt.Type {
-		// only concerned about PeerConnected, PeerFailedToConnect, and PeerAlreadyConnected
-		case PeerConnected, PeerFailedToConnect, PeerAlreadyConnected:
-		default:
-			return
-		}
 
-		// try to find a watcher for this peer
-		s.joinWatchersLock.Lock()
-		errCh, ok := s.joinWatchers[evnt.PeerID]
-		if ok {
-			errCh <- nil
-			delete(s.joinWatchers, evnt.PeerID)
-		}
-		s.joinWatchersLock.Unlock()
-	})
+	return p.Client(stream), nil
 }
 
-func (s *Server) Close() error {
-	err := s.host.Close()
-	s.dialQueue.Close()
-	close(s.closeCh)
+// getProtoStream returns an active protocol stream if present, otherwise
+// it returns nil
+func (s *Server) getProtoStream(protocol string, peerID peer.ID) *rawGrpc.ClientConn {
+	s.peersLock.Lock()
+	defer s.peersLock.Unlock()
 
-	return err
+	connectionInfo, ok := s.peers[peerID]
+	if !ok {
+		return nil
+	}
+
+	return connectionInfo.getProtocolStream(protocol)
 }
 
-func (s *Server) NewProtoStream(proto string, id peer.ID) (interface{}, error) {
-	s.protocolsLock.Lock()
-	defer s.protocolsLock.Unlock()
+// NewDiscoveryClient returns a new or existing discovery service client connection
+func (s *Server) NewDiscoveryClient(peerID peer.ID) (proto.DiscoveryClient, error) {
+	// Check if there is an active stream connection already
+	if protoStream := s.getProtoStream(common.DiscProto, peerID); protoStream != nil {
+		return proto.NewDiscoveryClient(protoStream), nil
+	}
 
-	p, ok := s.protocols[proto]
-	if !ok {
-		return nil, fmt.Errorf("protocol not found: %s", proto)
+	// Create a new stream connection and return it
+	protoStream, err := s.newProtoConnection(common.DiscProto, peerID)
+	if err != nil {
+		return nil, err
 	}
 
-	stream, err := s.NewStream(proto, id)
+	// Discovery protocol streams should be saved,
+	// since they are referenced later on
+	s.peersLock.Lock()
```

Comment:
This logic could go in a separate function

Reply:
Extracted 👍 

[633131d](https://github.com/0xPolygon/polygon-edge/pull/343/commits/633131dc63ac9f0106a71ccff5d4b8a0ab361fb5)